File: Ros/Day01/hw2/parkdoyeon_ros_rclcpp_pkg/src/publisher.cpp
#include "parkdoyeon_ros_rclcpp_pkg/publisher.hpp"

#include <chrono>
#include <functional>

using namespace std::chrono_literals;

PublisherNode::PublisherNode()
: Node("cpp_publisher"), count_(0)
{
    int_publisher_ =
        this->create_publisher<std_msgs::msg::Int32>(
            "cpp_int_topic", 10);

    string_publisher_ =
        this->create_publisher<std_msgs::msg::String>(
            "cpp_string_topic", 10);

    float_publisher_ =
        this->create_publisher<std_msgs::msg::Float32>(
            "cpp_float_topic", 10);

    bool_publisher_ =
        this->create_publisher<std_msgs::msg::Bool>(
            "cpp_bool_topic", 10);

    timer_ = this->create_wall_timer(
        1s,
        std::bind(&PublisherNode::timer_callback, this));
}

void PublisherNode::timer_callback()
{
    std_msgs::msg::Int32 int_msg;
    std_msgs::msg::String string_msg;
    std_msgs::msg::Float32 float_msg;
    std_msgs::msg::Bool bool_msg;

    int_msg.data = count_;
    string_msg.data = "Hello from C++";
    float_msg.data = 3.14f;
    bool_msg.data = (count_ % 2 == 0);

    int_publisher_->publish(int_msg);
    string_publisher_->publish(string_msg);
    float_publisher_->publish(float_msg);
    bool_publisher_->publish(bool_msg);

    RCLCPP_INFO(
        this->get_logger(),
        "int: %d, string: %s, float: %.2f, bool: %s",
        int_msg.data,
        string_msg.data.c_str(),
        float_msg.data,
        bool_msg.data ? "true" : "false");

    count_++;
}

int main(int argc, char * argv[])
{
    rclcpp::init(argc, argv);

    auto node = std::make_shared<PublisherNode>();

    rclcpp::spin(node);

    rclcpp::shutdown();

    return 0;
}

File: Ros/Day01/hw2/parkdoyeon_ros_rclpy_pkg/parkdoyeon_ros_rclpy_pkg/publisher.py
import rclpy

from rclpy.node import Node

from std_msgs.msg import Int32
from std_msgs.msg import String
from std_msgs.msg import Float32
from std_msgs.msg import Bool


class PublisherNode(Node):

    def __init__(self):
        super().__init__('py_publisher')

        self.int_publisher = self.create_publisher(
            Int32,
            'py_int_topic',
            10
        )

        self.string_publisher = self.create_publisher(
            String,
            'py_string_topic',
            10
        )

        self.float_publisher = self.create_publisher(
            Float32,
            'py_float_topic',
            10
        )

        self.bool_publisher = self.create_publisher(
            Bool,
            'py_bool_topic',
            10
        )

        self.timer = self.create_timer(
            1.0,
            self.timer_callback
        )

        self.count = 0

    def timer_callback(self):
        int_msg = Int32()
        string_msg = String()
        float_msg = Float32()
        bool_msg = Bool()

        int_msg.data = self.count
        string_msg.data = 'Hello from Python'
        float_msg.data = 2.71
        bool_msg.data = self.count % 2 == 0

        self.int_publisher.publish(int_msg)
        self.string_publisher.publish(string_msg)
        self.float_publisher.publish(float_msg)
        self.bool_publisher.publish(bool_msg)

        self.get_logger().info(
            f'int: {int_msg.data}, '
            f'string: {string_msg.data}, '
            f'float: {float_msg.data:.2f}, '
            f'bool: {bool_msg.data}'
        )

        self.count += 1


def main(args=None):
    rclpy.init(args=args)

    node = PublisherNode()

    rclpy.spin(node)

    node.destroy_node()

    rclpy.shutdown()


if __name__ == '__main__':
    main()